### PR TITLE
#21299: Fix slice_write channels%16!=0

### PR DESCRIPTION
**Pull request by sankarmanoj-tt**
### Ticket
#21299

### Problem description
- fast_compare in run_conv always passes, causing failing tests to appear as passing.
- Conv_Dram fails when out_channels%16!=0. When out_channels is not a multiple of 16, the conv op pads the out_channels dimension to make it a multiple. However, the final DRAM output has the actual out_channels dimension, not the padded one. Untilize With Unpad is used to remove this padding. However the output of Untilize with Unpad is incorrect.

### What's changed

- Change fast_compare function to an element-wise comparison which actually works
- Fixed slice write for out_channels%16!=0. Untilize is explicitly used to avoid untilize_with_unpad which fails. 
- Slice write does unpad on the channels dimension when it writes the output to DRAM Interleaved. 

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15302705128)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15302720490)
- [ ] Blackhole Nightly [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15302716059)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15304818647)
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15304824615)
- [x] Nightly L2 [passes](https://github.com/tenstorrent/tt-metal/actions/runs/15302712227) 
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp [hunk 1]
```diff
@@ -350,6 +350,7 @@ Result conv2d_DRAM(
             sliced_output_tensor = ttnn::to_memory_config(
                 sliced_output_tensor, MemoryConfig{TensorMemoryLayout::INTERLEAVED, BufferType::L1});
         }
+        sliced_output_tensor = ttnn::reshape(sliced_output_tensor, sliced_output_tensor.get_padded_shape());
```
Comment (ntarafdar) [inline ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:353]: can you clarify why this is padded shape and not the logical shape?

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:353]: This tensor has it's channels dimension padded, which needs to be unpadded. The slice_write op has been modified so that if the channels dimension of it's input is bigger than the channels dimension of the output, i.e. it's been padded, then it will do the unpad when it does the copy. 

To enable this, the tensors padded shape has been made it's logical shape.

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp [hunk 2]
```diff
@@ -41,7 +42,7 @@ void kernel_main() {
     constexpr bool dst0_is_dram = get_compile_time_arg_val(1) == 1;
 
     const InterleavedAddrGen<dst0_is_dram> s0 = {.bank_base_address = dst_addr, .page_size = padded_stick_size};
-
+    const uint32_t noc_write_size = min(padded_stick_size, unpadded_stick_size);
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp:45]: What is the scenario when the padded stick size is smaller than the unpadded one?

File: ttnn/cpp/ttnn/operations/experimental/slice_write/slice_write.cpp [hunk 3]
```diff
@@ -84,15 +86,32 @@ ttnn::Tensor SliceWriteOperation::invoke<uint32_t, 4>(
...
-        uint32_t input_nhw_volume = padded_input_shape[0] * padded_input_shape[1] * padded_input_shape[2];
         uint32_t calc_nhw_volume = actual_shape[0] * actual_shape[1] * actual_shape[2];
+        auto shard_spec = input_tensor.shard_spec().value();
+
+        auto input_cores = shard_spec.grid;
+        auto input_shard_shape = shard_spec.shape;
+        bool rm_orientation = shard_spec.orientation == ShardOrientation::ROW_MAJOR;
+        bool is_block_sharded = input_tensor.memory_config().memory_layout() == TensorMemoryLayout::BLOCK_SHARDED;
+        auto total_cores = shard_spec.grid;
+        uint32_t num_cores_nhw = total_cores.num_cores();
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/experimental/slice_write/slice_write.cpp:97]: Same thing if we have WIDTH sharding, is this correct?

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp [hunk 4]
```diff
@@ -189,10 +190,29 @@ SliceWriteRuntimeArgs get_slice_write_runtime_args_rm_sharded_input(
...
 
     TT_FATAL(input_tensor.shard_spec().has_value(), "Input tensor should be sharded");
+
     auto shard_spec = input_tensor.shard_spec().value();
     auto input_cores = shard_spec.grid;
     auto input_shard_shape = shard_spec.shape;
+
+    bool rm_orientation = shard_spec.orientation == ShardOrientation::ROW_MAJOR;
+    bool is_block_sharded = input_tensor.memory_config().memory_layout() == TensorMemoryLayout::BLOCK_SHARDED;
+    uint32_t num_cores_channels = 1;
```
Comment (dstoiljkovicTT) [inline ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp:200]: Do we support WIDTH sharding or only BLOCK and HEIGHT. If we do is this correct?

**Review by dstoiljkovicTT — COMMENTED** [review]
Could you please update the description with the implementation details, especially focusing on the padded shape becoming logical one?

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp (hunk 2 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp:45]: Padded stick size is smaller than unpadded stick size when the input's channels dimension has been padded. This happens when the conv has out_channels%16!=0. 

 If the input is block sharded, then the unpadded stick size gets divided by num_cores_channels. This makes unpadded_stick_size smaller than the padded_stick_size.

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp (hunk 4 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp:200]: Width sharded is not supported. Conv DRAM slicing is ideally for supporting large resolution inputs/outputs, which have large NHW. So width sharding will not be used. 
 
 If auto shard does select width sharding, there is a fall back by converting it to interleaved.

File: ttnn/cpp/ttnn/operations/experimental/slice_write/slice_write.cpp (hunk 3 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/experimental/slice_write/slice_write.cpp:97]: This op doesn't support width sharding.

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp [hunk 5]
```diff
@@ -16,6 +16,7 @@ void kernel_main() {
     const uint32_t num_sticks_per_core_read = get_arg_val<uint32_t>(7);
     const uint32_t num_read_per_barrier = get_arg_val<uint32_t>(8);
 
+#define DEBUG
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp:19]: Removed.

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp [hunk 6]
```diff
@@ -10,6 +10,7 @@
 #include <tt-metalium/host_api.hpp>
 
 #include "slice_write_op.hpp"
+#include "tt-metalium/assert.hpp"
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/experimental/slice_write/device/slice_write_program_factory.cpp]: It's used, but not necessary. VSCode adds these imports. I've removed them.

File: ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp [hunk 7]
```diff
@@ -16,6 +16,7 @@ void kernel_main() {
...
 #ifdef DEBUG
     DPRINT << "dst_addr: " << dst_addr << ENDL();
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/experimental/slice_write/device/kernels/dataflow/slice_write_writer_interleaved.cpp]: I would prefer to keep them for debugging in the future.

**Comment by dstoiljkovicTT** [comment]
Make sure to have CI pipelines passing but otherwise looks good to me.

**Comment by sankarmanoj-tt** [comment]
@s-jovic @ntarafdar  The reason for converting the logical shape to the padded shape was an issue with to_layout(Row Major). Because to_layout sees that the channels are padded, it calls UntilizeWithUnpadding which fails. By making the padded_shape the logical_shape, it becomes just untilize instead.

File: ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp (hunk 1 above)
Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:356]: https://github.com/tenstorrent/tt-metal/pull/22472#issuecomment-2909830264

Comment (sankarmanoj-tt) [reply ttnn/cpp/ttnn/operations/conv/conv2d/conv2d.cpp:356]: I have not debugged UntilizeWithUnpadding. But the output from the op is wrong.

File: tests/ttnn/nightly/unit_tests/operations/conv/test_conv2d.py [hunk 8]
```diff
@@ -3219,7 +3223,7 @@ def test_conv2d_sdxl(
         (1, 512, 512, 512, 512, ttnn.bfloat8_b, ttnn.bfloat16, 1, (3, 3), (1, 1), (1, 1), (1, 1), True, False, 1, 1, ttnn.Conv2dSliceWidth, 8),
 
         # output_channels 3
-        # (1, 128, 3, 1024, 1024, ttnn.bfloat8_b, ttnn.bfloat16, 1, (3, 3), (1, 1), (1, 1), (1, 1), True, False, 1, 1, ttnn.Conv2dSliceWidth, 16), #  pcc: 0.0
+        (1, 128, 3, 1024, 1024, ttnn.bfloat8_b, ttnn.bfloat16, 1, (3, 3), (1, 1), (1, 1), (1, 1), True, False, 1, 1, ttnn.Conv2dSliceWidth, 16), #  pcc: 0.0
```
[the remark this answers is not in the document]
Comment (sankarmanoj-tt) [reply tests/ttnn/nightly/unit_tests/operations/conv/test_conv2d.py]: Removed.

**Review by s-jovic — APPROVED** [review]
Please run pipelines with the latest changes before merging
